File: go/2840.go
package main

func checkStrings(s1 string, s2 string) bool {
	even := map[rune]int{}
	odd := map[rune]int{}

	for idx, ch := range s1 {
		if idx%2 == 0 {
			even[ch]++
		} else {
			odd[ch]++
		}
	}

	for idx, ch := range s2 {
		if idx%2 == 0 {
			ct, ok := even[ch]
			if !ok || ct <= 0 {
				return false
			}

			if ct == 1 {
				delete(even, ch)
			} else {
				even[ch]--
			}

		} else {
			ct, ok := odd[ch]
			if !ok || ct <= 0 {
				return false
			}

			if ct == 1 {
				delete(odd, ch)
			} else {
				odd[ch]--
			}

		}
	}

	return len(even) == 0 && len(odd) == 0
}

File: zig/2840.zig
const std = @import("std");

fn solution(s1: []const u8, s2: []const u8) bool {
    var even = [_]i32{0} ** 26;
    var odd = [_]i32{0} ** 26;

    for (s1, 0..) |ch, idx| {
        if (@mod(idx, 2) == 0) {
            even[ch - 'a'] += 1;
        } else {
            odd[ch - 'a'] += 1;
        }
    }

    for (s2, 0..) |ch, idx| {
        if (@mod(idx, 2) == 0) {
            even[ch - 'a'] -= 1;

            if (even[ch - 'a'] < 0) {
                return false;
            }
        } else {
            odd[ch - 'a'] -= 1;

            if (odd[ch - 'a'] < 0) {
                return false;
            }
        }
    }

    for (0..26) |idx| {
        if (even[idx] > 0 or odd[idx] > 1) {
            return false;
        }
    }

    return true;
}

const expect = std.testing.expectEqual;

test "test case #1" {
    const s1 = "abcdba";
    const s2 = "cabdab";

    const expected = true;
    const actual = solution(s1, s2);

    try expect(expected, actual);
}

test "test case #2" {
    const s1 = "abe";
    const s2 = "bea";

    const expected = false;
    const actual = solution(s1, s2);

    try expect(expected, actual);
}
